core/src/test/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerFacadeTest.java

```diff
@@ -0,0 +1,482 @@
+    public void whenPreparing_thenDoesNotReleaseMediaPlayer() {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer, never()).reset();
+        verify(mediaPlayer, never()).release();
+    }
+
+    @Test
+    public void whenPreparingMultipleTimes_thenReleasesMediaPlayer() {
+        facade.prepareVideo(ANY_URI);
+        facade.prepareVideo(ANY_URI);
+
+        verify(mediaPlayer).reset();
+        verify(mediaPlayer).release();
+    }
+
+    @Test
+    public void whenPreparing_thenSetsDataSource() throws IOException {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer).setDataSource(context, ANY_URI, NO_HEADERS);
+    }
+
+    @Test
+    public void whenPreparing_thenSetsDisplay() {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer).setDisplay(surfaceHolder);
+    }
+
+    @Test
+    public void whenPreparing_thenSetsStreamMusicAudioStreamType() {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer).setAudioStreamType(AudioManager.STREAM_MUSIC);
+    }
+
+    @Test
+    public void whenPreparing_thenSetsScreenOnWhilePlayerToTrue() {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer).setScreenOnWhilePlaying(SCREEN_ON);
+    }
+
+    @Test
+    public void whenPreparing_thenPreparesMediaPlayerAsyncronously() {
+        givenMediaPlayerIsPrepared();
+
+        verify(mediaPlayer).prepareAsync();
+    }
+
+    @Ignore("We should forward to listeners / make logs testable / throw exception.")
+    @Test
+    public void givenException_whenPreparing_thenLogsCreationError() {
+
+    }
+
+    @Test
+    public void givenBoundPreparedListener_andMediaPlayerIsPrepared_whenPrepared_thenForwardsOnPrepared() {
+        facade.prepareVideo(ANY_URI);
+        ArgumentCaptor<MediaPlayer.OnPreparedListener> argumentCaptor = ArgumentCaptor.forClass(MediaPlayer.OnPreparedListener.class);
+        verify(mediaPlayer).setOnPreparedListener(argumentCaptor.capture());
+        argumentCaptor.getValue().onPrepared(mediaPlayer);
+
+        verify(preparedListener).onPrepared(mediaPlayer);
+    }
+
+    @Ignore("Should we throw when a listener is not attached?")
+    @Test
+    public void givenNoBoundPreparedListener_andMediaPlayerIsPrepared_whenPrepared_thenDoesNotForwardOnPrepared() {
+
+    }
+
+    @Test
+    public void givenBoundVideoSizeChangedListener_andMediaPlayerOnPrepared_whenVideoSizeChanges_thenForwardsSizeChanges() {
+        givenMediaPlayerIsPrepared();
+
+        ArgumentCaptor<MediaPlayer.OnVideoSizeChangedListener> argumentCaptor = ArgumentCaptor.forClass(MediaPlayer.OnVideoSizeChangedListener.class);
+        verify(mediaPlayer).setOnVideoSizeChangedListener(argumentCaptor.capture());
+        argumentCaptor.getValue().onVideoSizeChanged(mediaPlayer, ANY_WIDTH, ANY_HEIGHT);
+
+        verify(videoSizeChangedListener).onVideoSizeChanged(eq(mediaPlayer), eq(ANY_WIDTH), eq(ANY_HEIGHT));
+    }
+
+    @Ignore("Should we throw when a listener is not attached?")
```

Comment:
This is not very well defined in `no-player`, `ExoPlayer` does something different to `AndroidMediaPlayer`